src/Servers.jl

```diff
@@ -139,46 +139,34 @@ closedorclosing(st) = st == CLOSING || st == CLOSED
 
 function requestclose!(c::Connection)
     if c.state == IDLE
-        closewriteandwait(c)
-        close(c)
         c.state = CLOSED
+        close(c)
     else
         c.state = CLOSING
     end
     return
 end
 
-function closewriteandwait(c::Connection)
-    io = IOExtras.tcpsocket(ConnectionPool.getrawstream(c))
-    @try begin
-        flush(io)
-        closewrite(io)
-        sleep(0.5) # give time for client to receive FIN
-    end
-    return
-end
-
 function closeconnection(c::Connection)
     c.state = CLOSED
-    closewriteandwait(c)
     close(c)
     return
 end
 
-# graceful shutdown that waits for active connectiosn to finish being handled
+# graceful shutdown that waits for active connections to finish being handled
 function Base.close(s::Server)
     shutdown(s.on_shutdown)
     close(s.listener)
     # first pass to mark or request connections to close
-    for c in s.connections
+    for c in collect(s.connections)
```

Comment:
why `collect` here?

Reply:
close is a yield point and you are missing a lock here. The collect helps minimize the data race window